src/persist-client/src/internal/apply.rs

```diff
@@ -83,68 +199,77 @@ where
         shared_states: &StateCache,
     ) -> Result<Self, Box<CodecMismatch>> {
         let shard_metrics = metrics.shards.shard(&shard_id);
-        let state = metrics
-            .cmds
-            .init_state
-            .run_cmd(&shard_metrics, |_cas_mismatch_metric| {
-                // No cas_mismatch retries because we just use the returned
-                // state on a mismatch.
-                state_versions.maybe_init_shard(&shard_metrics)
+        let state = shared_states
+            .get::<K, V, T, D, _, _>(shard_id, || {
+                metrics
+                    .cmds
+                    .init_state
+                    .run_cmd(&shard_metrics, |_cas_mismatch_metric| {
+                        // No cas_mismatch retries because we just use the returned
+                        // state on a mismatch.
+                        state_versions.maybe_init_shard(&shard_metrics)
+                    })
             })
             .await?;
-        // TODO: StateCache usage would look like the following.
-        if false {
-            let _state = shared_states
-                .get::<K, V, T, D, _, _>(shard_id, || {
-                    metrics
-                        .cmds
-                        .init_state
-                        .run_cmd(&shard_metrics, |_cas_mismatch_metric| {
-                            // No cas_mismatch retries because we just use the returned
-                            // state on a mismatch.
-                            state_versions.maybe_init_shard(&shard_metrics)
-                        })
-                })
-                .await?;
-        }
+        let cached_state = {
+            let state = state.lock().expect("lock poisoned");
+            CachedState::from(&*state)
+        };
         Ok(Applier {
             cfg,
             metrics,
             shard_metrics,
             state_versions,
             state,
+            cached_state,
         })
     }
 
-    // TODO: Remove usages of this.
-    pub fn state(&self) -> &TypedState<K, V, T, D> {
-        &self.state
+    pub fn state(&self) -> Arc<Mutex<TypedState<K, V, T, D>>> {
```

Comment:
for when you clean this up for review, IMO we shouldn't have this sort of method here and should replace uses of it with more targeted methods. the goal being that the Mutex never leaks out of `mod apply`, making it more local to verify claims that the lock is only held for short periods

Reply:
yeah I can move them, just seemed weird to ask the `Applier` for such details